text/0039-service-overhaul.md

```diff
@@ -0,0 +1,435 @@
+  - `void obs_service_enable_bandwidth_test(const obs_service_t *service, bool enabled)`: Enable/disable the service bandwidth test
+  - `bool obs_service_bandwidth_test_enabled(const obs_service_t *service)`; Return if the service bandwidth test is enabled
+  - `int obs_service_get_max_codec_bitrate(const obs_service_t *service, const char *codec)`: Return the maximum bitrate supported by the service depending on the codec
+  - `void obs_service_get_supported_resolutions2( const obs_service_t *service, struct obs_service_resolution **resolutions, size_t *count, bool *with_fps)`: Return resolutions  supported by the service with optional framerate
+  - `int obs_service_get_max_video_bitrate(const obs_service_t *service, const char *codec, struct obs_service_resolution resolution)`: Return the maximum bitrate supported by the service depending on the codec and a resolution
+  - `void obs_service_apply_encoder_settings2(obs_service_t *service, const char *encoder_id, obs_data_t *encoder_settings)`: Apply encoder settings for a specific encoder id
+
+
+Deprecating in the Services API:
+  - `obs_service_apply_encoder_settings`: replaced by `obs_service_apply_encoder_settings2` to allow per encoder id/codec settings
+  - `obs_service_get_supported_resolutions`: replaced by `obs_service_get_supported_resolutions2` to enable returning framerate
+  - `obs_service_get_max_bitrate`: replaced by `obs_service_get_max_codec_bitrate` and  `obs_service_get_max_video_bitrate` for per codec (and resolution for the second) bitrate
+
+### `rtmp-services`
+
+Services provided by this plugin (`"rtmp_custom"`, `"rtmp_common"`) will be deprecated (`OBS_SERVICE_DEPRECATED` and `OBS_SERVICE_INTERNAL` applied) and completely unused in OBS Studio.
+
+Those are deprecated rather than completely removed to allow scripting and plugins to migrate if they happen to use them.
+
+Its service JSON will no longer be updated.
+
+
+### `custom-services`
+
+If OBS Studio needs to be heavily dependent on one service plugin, it must be this one.
+
+This plugin is meant to provide replacements for the `"rtmp_custom"` type.
+
+#### `"custom_service"`
+
+This the only service of this plugin that will be exposed to the user through the UI.
+
+The protocol will be selected by the user and the properties view will change depending of the protocol.
+
+This service will support all first-party protocols, the properties view will change visible fields according to the protocol.
+
+#### Custom type per protocol
+
+Service types that support only the protocol inside their id (`"custom_`*insert lowercase protocol acronym*`"`) for any first-party protocol except FTL (deprecation).
+
+Services that will only support the first-party protocol related to their id/type.
+The flag `OBS_SERVICE_INTERNAL` will be applied to not show them in the UI list. Those are meant for being used in WebSocket and scripting.
```

Comment:
What does the migration path look like for scripts / apps using the old method?

Can we just provide a built in way to identify the old behavior and translate to the new one instead of keeping these around?

Reply:
`rtmp_common` and `rtmp_custom` are not removed but deprecated.
At some point we have to remove them.